### PR TITLE
issue #1060 pgp decrypt method that mimics error behavior

### LINKED ISSUE #1060: implement pgp decrypt method that mimics error behavior
part of #1051 , needs to be done before doing #1057

### PR DESCRIPTION
**Pull request by IvanPizhenko**
This PR introduces 100% jvm based method for the decryption.

close #1060
----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [ ] addresses the issue it closes (if any)
- [ ] code is readable and understandable
- [ ] is accompanied with tests, or tests are not needed
- [ ] is free of vulnerabilities

### CODE REVIEW

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (tomholub) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:182]: Each key may be protected by a different pass phrase. I think we could use some type like `KeyWithOptionalPassPhrase[]` when one key matches one possible pp.

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/kotlin/ByteArrayExt.kt
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/extensions/kotlin/ByteArrayExt.kt:25]: It seems that this method is unused in the code

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (tomholub) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:261]: This seems a naive way to do it - just because pass phrase was present does not mean this was an issue with the pass phrase. It could be a wrong key, for example.

Comment (tomholub) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:261]: Also it looks like the return errors should have one more item to be clearer, which would be `WRONG_PASS_PHRASE`

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/kotlin/ByteArrayExt.kt
Comment (DenBond7) [reply FlowCrypt/src/main/java/com/flowcrypt/email/extensions/kotlin/ByteArrayExt.kt:25]: Actually, all code in this file is unused. @IvanPizhenko Please clarify should we leave it as-is?

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (tomholub) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:274]: This is also naive - this could be a whole range of other issues.

**Review by tomholub — COMMENTED** [review]
Thank you - this is what I needed, but needs some improvements. I didn't look at the tests yet (will later).

It seems FORMAT error are not handled? For example, what will happen if I enter `"abcd".toByteArray()` as the data? It seems like it would throw "OTHER / wrong key" which is not correct.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:109]: Please look at this data class. It seems some of the fields can be null by default(for example `filename`). In that case we can a little shorter variant. For example

```
return DecryptionResult(
            isEncrypted = false,
            error = DecryptionError(type, message, cause)
        )
```

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:166]: Please clarify how I should understand this code? Maybe there is a bug.
`val armored: String? = null` -> it means `armored` can be a null
`val keyRing: PGPSecretKeyRing = readArmored(armored!!)` -> `armored!!` -> it means `armored` can not be a null

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:171]: I think we have PGPainless realization for this purpose

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (DenBond7) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:179]: It seems this method uses `ByteArray` as an input. 

And here we use RAM twice.

```
val output = ByteArrayOutputStream()
decryptionStream.use { it.copyTo(output) }
```

@tomholub I think the current approach is enough to drop Node.js (I mean don't care about RAM). But it has performance issues for Android. We will have to improve that in the near future(upcoming releases).

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (DenBond7) [inline FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:308]: @tomholub is it safe to store email addresses here? as I understand it's a real email address

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (DenBond7) [inline FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:28]: @tomholub @IvanPizhenko @seisvelas I'm thinking about the code style here. As for me, it's not so good to store such a long string here. It bloats the code and makes it less readable. I don't need to know the whole content of a private key or an encrypted message to analyze tests. Instead, now I have to do a lot of scrolling. In my opinion, it'll be better to store such data in resources.

**Review by DenBond7 — CHANGES_REQUESTED** [review]
Leaved a few comments. Please look at them.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (tomholub) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:179]: That's ok, optimizations can be done in a follow-up. Thanks for mentioning - you can file an issue with some things to look into afterwards.

Comment (DenBond7) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:179]: > And here we use RAM twice.

Sorry, not twice. Just more.

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (tomholub) [reply FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:308]: For the purposes of these tests on Android, we should not be storing it urlencoded this way. @IvanPizhenko please extract and url-decode the armored message from each of these strings, and put that in the tests instead. Since the message is what you're actually testing, the rest is not meaningful here.

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (tomholub) [reply FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:28]: I have no preference either way. By putting them here, you have to scroll but it will never get out of sync. Maybe by putting in resources you could later delete the test but leave the resource still there? So we may accumulate resources that are unused. But it makes tests indeed cleaner and more readable.

On Android, we should just follow whatever you have already been doing in other tests.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpArmor.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpArmor.kt:166]: yes, maybe it is worth for inclusion into PGPainless.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:182]: Each key or each key in the keyring? PGPainless does not handle such case as somethings standard. If that is the case, maybe it is worth to implement in the PGPainless. Also as you can see, we only support only one keyring now, so such data structure may be postponed.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:261]: There is no absolutely precise way to tell what has happened. We only can guess. So the guess is if we could not restore session key for some reason and there is passphrase, the cause is likely wrong password, otherwise wrong key. This is similar to what Typescript does.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:274]: will change to something more general like "decryption failed"

File: FlowCrypt/src/main/java/com/flowcrypt/email/extensions/kotlin/ByteArrayExt.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/extensions/kotlin/ByteArrayExt.kt:25]: Let's remove it

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:109]: yes, can be made better, will update it

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:171]: I will check.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (tomholub) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:261]: As far as I can tell, TypeScript we know with certainty which situation occurred.

If we cannot tell here, then we need to improve PGPainless so that it throws clearer errors.

But for pass phrase - and this also applies to https://github.com/FlowCrypt/flowcrypt-android/pull/1216#discussion_r628691330 - until we include that functionality in PGPainless, we could simply decrypt each key ahead of time before passing into PGPainless.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (tomholub) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:182]: Sorry, I'm not too familiar with Java PGP terminology. Here is how I'm using the words:

key packet - An OpenPGP key packet like primary key packet or subkey packet
OpenPGP key - a collection of usually one master key packet and some number of sub-packets

I don't know what a keyring is, but presumably it's a collection of OpenPGP keys.

Each OpenPGP key may have a different pass phrase. Whether these are represented during decryption as one key ring, or several key rings, or several OpenPGP keys differently is an implementation detail (which we may need to resolve).

Technically, each key packet within OpenPGP key may have a different pass phrase, but we don't support this scenario (and check for it during setup), so you will not encounter such keys on Android.

Now, for this actual issue, you may use a temporary solution as in https://github.com/FlowCrypt/flowcrypt-android/pull/1216#discussion_r628708349 (decrypt all yourself first) and then later we include this functionality in PGPainless, so that we don't have to decrypt keys and packets that are not used for a particular message.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:166]: I think `armored` should be non-nullable, I will fix this

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:179]: Let's do it later. Added #1219

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:308]: @tomholub I am not disclosing here anything that is not already disclosed. These tests come from flowcrypt-mobile-core repo, which is also public.

Comment (IvanPizhenko) [reply FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:308]: But ok, will try to remove extra urlencoded stuff

Comment (tomholub) [reply FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:308]: I'll make an issue to address it there too

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:28]: will move to resources

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:182]: @tomholub BouncyCastle `PGPKeyRing` (and its descendants `PGPPublicKeyRing` and `PGPSecretKeyRing`, which you can see here in the code) matches to what you call "OpenPGP key". Keeping this in mind, we can say that current decryption routine supports only one "OpenPGP key". Should we pass there multiple ones? And then decrypt them and then pass decrypted ones to the PGPainess?

Comment (tomholub) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:182]: Thank you for explanation. While most users will have one OpenPGP key, they may also have more than one. Therefore we'll have to support supplying more than one OpenPGP key (or PGPSecretKeyRing) into the decrypt function.

As you say, we would decrypt them all first, and then supply into PGPainless decrypted.

In the future, we have to improve PGPainless to:
 - accept more than one PGPSecretKeyRing (if it doesn't already)
 - allow for different PGPSecretKeyRing to be unlocked by different pass phrase

But that's an optimization, after we drop NodeJS on Android.

> Should we pass there multiple ones? And then decrypt them and then pass decrypted ones to the PGPainess?

:+1:

Comment (tomholub) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:182]: > accept more than one PGPSecretKeyRing (if it doesn't already)

I think PGPainless already did that - correct me if I'm wrong. If it already supports passing several, then this will be easier. If not, they you'd need to call it in a loop, key by key..

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpArmor.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpArmor.kt:166]: I've posted https://github.com/pgpainless/pgpainless/issues/116

Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpArmor.kt:166]: For now I suggest to leave as is, and update to PGPainless later.

**Comment by IvanPizhenko** [comment]
@tomholub @DenBond7 made some fixes here, please review

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpArmor.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpArmor.kt:167]: Paul answered that PGPainless doesn't support this, so we can accept current implementation for now and update to PGPainless when Paul do it.

Comment (tomholub) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpArmor.kt:167]: Sounds good

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (tomholub) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:66]: Maybe `fourBlockTypes` is not the best name :-)

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (DenBond7) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:101]: :+1:

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (tomholub) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:192]: I think you meant to write `if(!KeyRingInfo(it.keyRing).isFullyDecrypted)`

However I would invert the logic - if it's already decrypted, then no need to worry about pass phrase.

```kt
if(KeyRingInfo(it.keyRing).isFullyDecrypted) {
  it.keyRing
} else if(it.passphrase == null) {
  throw PGPException("missing pass phrase to decrypt key")
} else {
  PgpKey.decryptKey(it.keyRing, it.passphrase) // may throw PGPException
}
```

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (tomholub) [inline FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:267]: Exception handling looks better now - over time we can fine tune it based on real world errors

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (tomholub) [inline FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:168]: Highlighting to file an issue for later

**Review by DenBond7 — APPROVED** [review]
Thank you. Looks good. Approved

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (tomholub) [inline FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:180]: This is a missing armor checksum - different from MDC. Usually the four digits at the and like `p3Fc=`. Such messages are still valid if this is missing, and should decrypt correctly - so it's good as is. Maybe you should add this comment into code so it doesn't confuse us later.

**Comment by DenBond7** [comment]
@tomholub Please review it. I will merge it after your approving

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (tomholub) [inline FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:161]: Is there a test somewhere to test:
 - wrong key provided
 - wrong pass phrase provided
 - key missing pass phrase

Also, please name the existing test more appropriately alike `missingMdcProducesError` and `modifiedMdcProducesError`

**Review by tomholub — COMMENTED** [review]
Looks good, with tiny changes, from my end it's good to merge it. I'll let Den do the final review and merging. Thank you!

**Comment by tomholub** [comment]
Ok Den, after the above changes, I'll do my approving review, then you can do yours.

File: FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt:192]: ok, will change this

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:168]: we already have https://github.com/FlowCrypt/flowcrypt-android/issues/1214 for this

Comment (IvanPizhenko) [reply FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:168]: also added https://github.com/FlowCrypt/flowcrypt-android/issues/1222 for another such question

File: FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt
Comment (IvanPizhenko) [reply FlowCrypt/src/test/java/com/flowcrypt/email/security/pgp/PgpMsgTest.kt:180]: Added comment

**Review by tomholub — APPROVED** [review]
Thank you! Looks good to me. Will let Den take another look and merge it.

**Review by DenBond7 — APPROVED** [review]

**Pull request merged**
